### ISSUE TITLE
Pass parameters to templates in aws-api-gateway-*

### ISSUE THREAD

**Issue by Gamecock** [awslabs/aws-solutions-constructs#891]
Based on [aws-api_gateway-dynamodb](https://github.com/awslabs/aws-solutions-constructs/tree/main/source/patterns/%40aws-solutions-constructs/aws-apigateway-dynamodb)

The constructor has read_request_template etc.

The only way I can make it work in the python SDK is hard code the table name.

I can see the default template [index.ts](https://github.com/awslabs/aws-solutions-constructs/blob/main/source/patterns/%40aws-solutions-constructs/aws-apigateway-dynamodb/lib/index.ts) line 268.

     const readRequestTemplate = props.readRequestTemplate ??
        `{ \
          "TableName": "${this.dynamoTable.tableName}", \
          "KeyConditionExpression": "${partitionKeyName} = :v1", \
          "ExpressionAttributeValues": { \
            ":v1": { \
              "S": "$input.params('${partitionKeyName}')" \
            } \
          } \
        }`;

I cannot figure out how to pass parameters into the template.

the create request template looks like it explicitly replaces {$Table}

**Comment by biffgaut** [comment]
While we recommend allowing the CDK/CloudFormation to create machine generated names for resources, as that allows multiple stacks to be created within 1 account/1 region, we see two ways to accomplish what you seek:

* Pass a table name in with the `dynamoTableProps`
* Submit an entire custom readRequestTemplate

**Comment by biffgaut** [comment]
Does [this PR](https://github.com/awslabs/aws-solutions-constructs/pull/898) provide what you need? It's merged and should be released this week.

**Comment by Gamecock** [comment]
I don't think it does, but I will try once it's released.  I'm not 100% sure if this is a bug or operator error.  If the latter once I figure it out I can update documentation.  This example is in the test file in V2.31.0

```
new ApiGatewayToDynamoDB(stack, 'test-api-gateway-dynamodb-additional-request-templates', {
  existingTableObj,
  additionalReadRequestTemplates: {
    'text/plain': `{ \
      "TableName": "${existingTableObj.tableName}", \
      "KeyConditionExpression": "${partitionKeyName} = :v1", \
      "ExpressionAttributeValues": { \
        ":v1": { \
          "S": "$input.params('${partitionKeyName}')" \
        } \
      } \
    }`
  }
});
```

I can get the `Stack.template.json` to include `\"${existingTableObj.tableName}\"` or `\"${existing_table_obj.table_name}\"` like the expected.json on a `cdk synth` but if I deploy and test the API Method, I get `""` for the "TableName".

`Wed Mar 01 15:50:50 UTC 2023 : Endpoint request body after transformations: { "TableName": "",  "KeyConditionExpression": "device = :v1"`

**Comment by Gamecock** [comment]
Sorry for the delay, but I had not worked with typescript before.  I was able to deploy, but could not confirm the additional request was working, as I could not tell it apart from the default template.  

My Goal is to be able to use a sort key and not use 'id' as the primary key.

This works in Typescript
import { App, Stack, StackProps, Duration } from 'aws-cdk-lib';
import { ApiGatewayToDynamoDBProps, ApiGatewayToDynamoDB } from "@aws-solutions-constructs/aws-apigateway-dynamodb";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import * as api_gateway from "aws-cdk-lib/aws-apigateway"

export class ApiGatewayTsStack extends Stack {
  constructor(scope: App, id: string, props?: StackProps) {
    super(scope, id, props);
    const partitionKeyName = 'ident';
    const sortKeyName = 'timelike'

    const existingTableObj = new dynamodb.Table(this, 'existing-table', {
      partitionKey: {
        name: partitionKeyName,
        type: dynamodb.AttributeType.STRING,
      },
      sortKey: {
        name: sortKeyName,
        type: dynamodb.AttributeType.NUMBER
      },
      pointInTimeRecovery: true,
      encryption: dynamodb.TableEncryption.AWS_MANAGED,
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST
    });

    new ApiGatewayToDynamoDB(this, 'test-api-gateway-dynamodb-additional-request-templates', {  existingTableObj,
      readRequestTemplate:  `{ \
          "TableName": "${existingTableObj.tableName}", \
          "KeyConditionExpression": "${partitionKeyName} = :v1  AND ${sortKeyName} < :v2", \
          "ExpressionAttributeValues": { \
            ":v1": { \
              "S": "$input.params('${partitionKeyName}')" \
            }, \
            ":v2": { "N": "$input.params('${sortKeyName}')"}
          } \
        }`
      }
    );


  }
}


This does not work in Python:

`from aws_solutions_constructs.aws_apigateway_dynamodb import ApiGatewayToDynamoDB, ApiGatewayToDynamoDBProps
from aws_cdk import Stack, aws_dynamodb as ddb, aws_apigateway
from constructs import Construct

class ApiGatewayBugStack(Stack):

    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        partitionKeyName = 'ident'
        sortKeyName = "timelike"

        existingTableObj = ddb.Table(self, 'existing-table',   
            partition_key = ddb.Attribute(
                name= partitionKeyName,
                type= ddb.AttributeType.STRING,
            ),
            sort_key= ddb.Attribute(
                name=sortKeyName,
                type= ddb.AttributeType.NUMBER
            ),
            point_in_time_recovery= True,
            encryption= ddb.TableEncryption.AWS_MANAGED,
            billing_mode= ddb.BillingMode.PAY_PER_REQUEST
            )

        ApiGatewayToDynamoDB(self, 'test-api-gateway-dynamodb-additional-request-templates', existing_table_obj=existingTableObj,
                api_gateway_props= aws_apigateway.RestApiProps(rest_api_name="Python_API"),
                read_request_template= repr({ \
                "TableName": "${existingTableObj.tableName}", \
                "KeyConditionExpression": "${partitionKeyName} = :v1 AND ${sortKeyName} < :v2", \
                "ExpressionAttributeValues": { \
                    ":v1": { \
                    "S": "$input.params('${partitionKeyName}')" \
                    }, \
                    "v2:" : { "N": "$input.params('${partitionKeyName}')" }, \
                } \
                })
            
        )
        `

The cloud formation template.json seems to be OK:

`     "RequestTemplates": {
      "application/json": "{'TableName': '${existingTableObj.tableName}', 'KeyConditionExpression': '${partitionKeyName} = :v1 AND ${sortKeyName} < :v2', 'ExpressionAttributeValues': {':v1': {'S': \"$input.params('${partitionKeyName}')\"}, 'v2:': {'N': \"$input.params('${partitionKeyName}')\"}}}"
     },`

But when it's executed in API Gateway test page, the name is not resolved:

Response:
`{"__type":"com.amazon.coral.service#SerializationException"}`

Logs Snippet:
`Tue Mar 21 18:49:53 UTC 2023 : Endpoint request body after transformations: {'TableName': '', 'KeyConditionExpression': ' = :v1 AND  < :v2', 'ExpressionAttributeValues': {':v1': {'S': ""}, 'v2:': {'N': ""}}}`

I have not tested any other languages.

**Comment by Gamecock** [comment]
I'd like for it to work with table props  vs existing table, but that is just a nice to have.  This method will work.
